### DISCUSSION TITLE
Rough estimate on when lifetimes, traits, and algebraic data types will be released?

### DISCUSSION THREAD

**Discussion by rarebreed** [modular/modular#1311] [Q&A] [answered]
If a rough estimate is hard to do, how about providing:

- a list of the prioritization of mojo development efforts
- a road map/checklist of items that are still being worked on for various features

I have been holding off on truly investing a lot of time learning mojo, due to the lack of the features mentioned in the title.  From my experience with rust, the ownership and lifetime concepts are a non-trivial change and forces rethinking how coding is done.  In effect, mojo with lifetimes + ownership will be a different language in many ways (even if mojo can do way more lifetime ellision than rust can, I think it is something mojicians will need to be aware of).  The syntax changes will be relatively minor I think, but the change in how you have to think about your code will be profound.  Compound this with proposals over what to do with keywords like `let`, `var`, `ref` and `mut` make me hesitant to do further exploration of mojo.

In fact, I am a little concerned that people may get used to programming mojo directly with raw Pointer types which is inherently unsafe (and brings up another topic of whether mojo will allow unsafe programming with raw Pointer types or at least how to declare code is using raw pointers). 

It'd be really cool if there was a way to play with prototype releases or have an experimental feature flag so we can try out developmental features.

**Comment by lattner** [comment, marked as answer]
I'd guess ~2-3 months for these features to get built out.  Features like traits will have many subfeatures though.

**Comment by lattner** [comment]
Very basic traits = yesterday!

**Comment by lattner** [comment]
I'm going to close this out to tidy up the discussion list
